src/hardware/driver/output.rs

```diff
@@ -231,19 +209,66 @@ where
                 if self.context().iir.y_offset
                     >= (*target - Output::<I2C>::RAMP_STEP)
                 {
+                    debug!("Current ramp done. Output enabled.");
                     self.process_event(sm::Events::RampDone).unwrap();
                 } else {
                     self.process_event(sm::Events::Tick).unwrap();
                 }
             }
+            sm::States::SelftestZero => {
+                tests = Some((
+                    [
+                        Output::<I2C>::VALID_VOLTAGE_ZERO,
+                        Output::<I2C>::VALID_CURRENT_ZERO,
+                    ],
+                    [FailReason::ZeroVoltage, FailReason::ZeroCurrent],
+                ));
+                self.process_event(sm::Events::Tick).unwrap();
+            }
+            sm::States::SelftestShunt => {
+                tests = Some((
+                    [
+                        Output::<I2C>::VALID_VOLTAGE_SHUNT,
+                        Output::<I2C>::VALID_CURRENT_SHUNT,
+                    ],
+                    [FailReason::ShuntVoltage, FailReason::ShuntCurrent],
+                ));
+                self.process_event(sm::Events::Tick).unwrap();
+            }
+            sm::States::SelftestShort => {
+                tests = Some((
+                    [
+                        Output::<I2C>::VALID_VOLTAGE_SHORT,
+                        Output::<I2C>::VALID_CURRENT_SHORT,
+                    ],
+                    [FailReason::ShortVoltage, FailReason::ShortCurrent],
+                ));
+                self.process_event(sm::Events::Tick).unwrap();
+            }
             _ => (),
         };
+
+        if let Some(tests) = tests {
+            for ((range, &read), &reason) in
+                tests.0.iter().zip(reads.iter()).zip(tests.1.iter())
+            {
+                if !range.contains(&read) {
+                    interlock.set(Some(Reason::Selftest(Selftest {
+                        reason,
+                        read,
+                        channel,
+                    })));
```

Comment:
If you want to reduce function size (and you should), this and a match for the range constants could have probably gone into the SelfTest enum impl. But also fine as it is.